pkg/ucp/store/postgres/postgresclient.go

```diff
@@ -0,0 +1,414 @@
+	if ctx == nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+
+	parsed, err := resources.Parse(id)
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must refer to a named resource, not a collection"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return nil, err
+	}
+
+	obj := store.Object{}
+	err = p.api.QueryRow(
+		ctx,
+		"SELECT original_id, etag, resource_data FROM resources WHERE id = $1",
+		storeutil.NormalizePart(normalized.String())).Scan(&obj.ID, &obj.ETag, &obj.Data)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return nil, &store.ErrNotFound{ID: id}
+	} else if err != nil {
+		return nil, err
+	}
+
+	return &obj, nil
+}
+
+// Query implements store.StorageClient.
+func (p *PostgresClient) Query(ctx context.Context, query store.Query, options ...store.QueryOptions) (*store.ObjectQueryResult, error) {
+	if ctx == nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+
+	err := query.Validate()
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: fmt.Sprintf("invalid argument. Query is invalid: %s", err.Error())}
+	}
+
+	config := store.NewQueryConfig(options...)
+
+	// For a scope query, we need to perform the same normalization as we do for other operations on scopes.
+	resourceType := storeutil.NormalizePart(query.ResourceType)
+	if query.IsScopeQuery {
+		var err error
+		resourceType, err = storeutil.NormalizeResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
+
+		resourceType = storeutil.NormalizePart(resourceType)
+	}
+
+	var routingScopePrefixFilter *string
+	if query.RoutingScopePrefix != "" {
+		routingScopePrefixFilter = to.Ptr(storeutil.NormalizePart(query.RoutingScopePrefix))
+	}
+
+	var timestampFilter *string
+	if config.PaginationToken != "" {
+		ts, err := p.parsePaginationToken(config.PaginationToken)
+		if err != nil {
+			return nil, &store.ErrInvalid{Message: "invalid argument. 'query.PaginationToken' is invalid."}
+		}
+		timestampFilter = &ts
+	}
+
+	var limitFilter *int
+	if config.MaxQueryItemCount > 0 {
+		limitFilter = &config.MaxQueryItemCount
+	}
+
+	// For a scope query, we need to perform the same normalization as we do for other operations on scopes.
+	if query.IsScopeQuery {
+		var err error
+		query.ResourceType, err = storeutil.NormalizeResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// NOTE: building SQL by concatenating strings is hard to do safely and should be avoided.
+	// If you need to work on this code MAKE SURE you use SQL parameters
+	// for any user input.
```

Comment:
Agreed! 👍

Reply:
FYI for reviewers. `Query.Filters` is executed on the client side rather than pushed to the server. I did this for compatibility with the existing code, because the testing done for `Query.Filters` is minimal. 

This comment is relevant because we **will** need to build SQL dynamically to do that 😢.